File: programmers/java/모음사전.java
package programmers.java;

public class 모음사전 {

    public static void main(String[] args) {
        모음사전_Solution solution = new 모음사전_Solution();
        int ans = solution.solution("AAAE");
        System.out.println(ans);
    }
}
class 모음사전_Solution {
    int answer = 0;
    String[] alphabet = {"A", "E", "I", "O", "U"};
    public int solution(String word) {
        dfs("", word);
        return answer-1;
    }

    private boolean dfs(String str, String word) {
        answer++;
        if (str.equals(word)) {
            return true;
        }
        if (str.length() == 5) {
            return false;
        }

        for (int i=0; i<5; i++) {
            if (dfs(str.concat(alphabet[i]), word)) {
                return true;
            }
        }
        return false;
    }
}

File: programmers/python/모음사전.py
def solution(word):
    alphabet = ['A', 'E', 'I', 'O', 'U']
    global answer
    answer = 0

    def bt(string):
        global answer
        answer += 1

        if word == string:
            return True
        if len(string) == 5:
            return False

        for i in range(5):
            if bt(string+alphabet[i]):
                return True
    for i in range(5):
        if bt(alphabet[i]):
            return answer


def solution2(word):
    alphabet = ['A', 'E', 'I', 'O', 'U']
    global answer
    answer = 0
    log = []

    def bt():
        global answer
        answer += 1

        if "".join(log) == word:
            return True
        if len(log) == 5:
            return False

        for i in range(5):
            log.append(alphabet[i])
            ret = bt()
            log.pop()
            if ret:
                return True
    bt()
    return answer-1


print(solution2("AAAAE"))
print(solution2("AAAE"))
print(solution2("I"))
print(solution2("EIO"))
